### PR TITLE
Add service generator

### PR DESCRIPTION
**Pull request by citizen428**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

I'm currently a bit blocked on profile work, waiting for 2 PRs to get approved and merged. So in the meantime I did some work on a service generator to standardize our service objects a bit more.

### Examples:

**Generate a non-namespaced service without arguments**

`$ rails generate service DoTheThing`

```ruby
# app/services/do_the_thing.rb
class DoTheThing
  def self.call
    new.call
  end

  def call
  end
end
```

```ruby
# spec/services/do_the_thing_spec.rb
require "rails_helper"

RSpec.describe DoTheThing, type: :service do
  pending "add some examples to (or delete) #{__FILE__}"
end
```

**Generate a non-namespaced service with arguments:**

`$ rails generate service DoTheThing arg1 arg2`

```ruby
# app/services/do_the_thing.rb
class DoTheThing
  def self.call(arg1, arg2)
    new(arg1, arg2).call
  end

  def initialize(arg1, arg2)
    @arg1 = arg1
    @arg2 = arg2
  end

  def call
  end
end
```

The generated spec is the same as above.

**Generate a namespaced service with arguments**

`$ rails generate service things/dothem arg1 arg2`

```ruby
# app/services/things/do_them.rb
class Things::DoThem
  def self.call(arg1, arg2)
    new(arg1, arg2).call
  end

  def initialize(arg1, arg2)
    @arg1 = arg1
    @arg2 = arg2
  end

  def call
  end
end
```

```ruby
# spec/services/things/do_them_spec.rb
require "rails_helper"

RSpec.describe Things::DoThem, type: :service do
  pending "add some examples to (or delete) #{__FILE__}"
end
```

## Related Tickets & Documents

n/a

## QA Instructions, Screenshots, Recordings

Try the different examples above and verify that the generated code matches your expectations.

## Added tests?

- [ ] Yes
- [X] No, and this is why: there's surprisingly little documentation around testing generators with RSpec. The examples I found all use an external gem for this. I'm not sure I want to add this dependency just yet and we are also not testing the other generator. As this is a convenience tool for developers and not production code I think this is ok until I can evaluate testing strategies during the next cooldown.
- [ ] I need help with writing tests

## Added to documentation?

- [X] Docs.forem.com
- [ ] README
- [ ] No documentation needed

### CODE REVIEW

**Review by rhymes — APPROVED** [review]
Cool!

**Comment by rhymes** [comment]
@citizen428 it's failing because of a naming issue:

> /home/travis/build/forem/forem/vendor/bundle/ruby/2.7.0/gems/zeitwerk-2.4.0/lib/zeitwerk/loader/callbacks.rb:17:in `on_file_autoloaded': expected file /home/travis/build/forem/forem/lib/generators/service/service_generator.rb to define constant Generators::Service::ServiceGenerator, but didn't (Zeitwerk::NameError)

**Comment by citizen428** [comment]
@rhymes You caught me during lunch, I had already started working on this. The error you posted is a result of an attempt to fix it. Before it failed with this:

```
/home/travis/build/forem/forem/lib/generators/service/service_generator.rb:1:in `<main>': uninitialized constant Rails::Generators (NameError)
```

I was a bit dumbfounded, as we do everything the same way as in the data update generator. However, some googling seems to suggest that explicitly requiring `rails/generators` is the ThingToDo™ since Rails 3:

* heartcombo/devise#2743
* thoughtbot/refills#33
* https://github.com/nsarno/knock/issues/156
* https://stackoverflow.com/questions/14883897/uninitialized-constant-railsgenerators-nameerror/14884631#14884631

But this wasn't necessary for the `DataUpdateGenerator`... Anyway, adding it does lead to the error you mentioned above, which makes sense with how Zeitwerk loads constants.

**Comment by citizen428** [comment]
This made me believe we manually updated the autoload path for the other generator and that's indeed the case:

```ruby
# config/initializers/zeitwerk.rb
# Ignoring folders that don't adhere to the new naming conventions
Rails.autoloaders.main.ignore(Rails.root.join("lib/data_update_scripts"))
Rails.autoloaders.main.ignore(Rails.root.join("lib/generators/data_update"))
```

I added a line for the new generator now.

**Comment by citizen428** [comment]
Oops, accidentally merge `master` back here instead of some other branch...

File: docs/backend/service-objects.md [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    @arg1 = arg1
+  end
+
+  def call
+    # import code goes here
+  end
+end
+```
+
+To distinguish services from models we often give them verb names vs nound
```
Comment (Ridhwana) [inline docs/backend/service-objects.md:30]:
```suggestion
To distinguish services from models we often give them verb names vs noun
```

**Review by Ridhwana — APPROVED** [review]
Looks good, I just suggested a change for what looks like a typo.

**Review by mstruve — APPROVED** [review]
🔥

File: docs/backend/service-objects.md (hunk 1 above)
Comment (citizen428) [reply docs/backend/service-objects.md]: Thanks @Ridhwana

**Pull request merged**